NBA teams are increasingly trotting out lineups with five players who can play and guard nearly any position. traditional positions don't accurately explain what a players skillset truly is, they incorrectly oversimplify the skill sets of NBA players. Simply plugging players into one of five positions does not accurately define a player’s specific skill set. Moreover, the misclassification of a player’s position may lead teams to waste resources on developing draft picks that do not fit their systems.

In light of these changes, we need an effective way to designate positions in the NBA not based on basic physical traits such as height and weight, but in terms of function, such as shooting and defense. A framework for modern NBA positions is important towards our understanding for how players have evolved, and effective roster construction. 

### Import the Required Python Packages and Methods

In [134]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import csv


### Finding and Evaluating Historical NBA Player Data

I will use the Kaggle dataset "NBA Players stats since 1950", with stats for all players since 1950. The file Seasons_Stats.csv contains the statics of all players since 1950, with details of each column as below:

    Year - Season
    Player - name
    Pos - Position
    Age - Age
    Tm - Team
    G - Games
    GS - Games Started
    MP - Minutes Played
    PER - Player Efficiency Rating
    TS% - True Shooting %
    3PAr - 3-Point Attempt Rate
    FTr - Free Throw Rate
    ORB% - Offensive Rebound Percentage
    DRB% - Defensive Rebound Percentage
    TRB% - Total Rebound Percentage
    AST% - Assist Percentage
    STL% - Steal Percentage
    BLK% - Block Percentage
    TOV% - Turnover Percentage
    USG% - Usage Percentage
    blanl
    OWS - Offensive Win Shares
    DWS - Defensive Win Shares
    WS - Win Shares
    WS/48 - Win Shares Per 48 Minutes
    blank2
    OBPM - Offensive Box Plus/Minus
    DBPM - Defensive Box Plus/Minus
    BPM - Box Plus/Minus
    VORP - Value Over Replacement
    FG - Field Goals
    FGA - Field Goal Attempts
    FG% - Field Goal Percentage
    3P - 3-Point Field Goals
    3PA - 3-Point Field Goal Attempts
    3P% - 3-Point Field Goal Percentage
    2P - 2-Point Field Goals
    2PA - 2-Point Field Goal Attempts
    2P% - 2-Point Field Goal Percentage
    eFG% - Effective Field Goal Percentage
    FT - Free Throws
    FTA - Free Throw Attempts
    FT% - Free Throw Percentage
    ORB - Offensive Rebounds
    DRB - Defensive Rebounds
    TRB - Total Rebounds
    AST - Assists
    STL - Steals
    BLK - Blocks
    TOV - Turnovers
    PF - Personal Fouls
    PTS - Points

In [135]:
with open('data/Seasons_Stats.csv','rb') as csvfile:
    reader1 = csv.reader(csvfile)
df1 = pd.read_csv('data/Seasons_Stats.csv')

First, we remove duplicated or empty rows and drop a couple of blank columns.

In [136]:
df1_no_duplicates = df1.drop_duplicates()
df2 = df1_no_duplicates
del df2['blanl']
del df2['blank2']
df2.dropna(how='all')
df2.head(5)

Unnamed: 0    Year           Player  Pos   Age   Tm     G  GS  MP  PER  \
0           0  1950.0  Curly Armstrong  G-F  31.0  FTW  63.0 NaN NaN  NaN   
1           1  1950.0     Cliff Barker   SG  29.0  INO  49.0 NaN NaN  NaN   
2           2  1950.0    Leo Barnhorst   SF  25.0  CHS  67.0 NaN NaN  NaN   
3           3  1950.0       Ed Bartels    F  24.0  TOT  15.0 NaN NaN  NaN   
4           4  1950.0       Ed Bartels    F  24.0  DNN  13.0 NaN NaN  NaN   

     TS%  3PAr    FTr  ORB%  DRB%  TRB%  AST%  STL%  BLK%  TOV%  USG%  OWS  \
0  0.368   NaN  0.467   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN -0.1   
1  0.435   NaN  0.387   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  1.6   
2  0.394   NaN  0.259   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  0.9   
3  0.312   NaN  0.395   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN -0.5   
4  0.308   NaN  0.378   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN -0.5   

   DWS   WS  WS/48  OBPM  DBPM  BPM  VORP     FG    FGA    FG%  3P  3PA  3P%  \
0  3.6  3.5    NaN   NaN   NaN  NaN   NaN  144.0  516.0  0.279 NaN  NaN  NaN   
1  0.6  2.2    NaN   NaN   NaN  NaN   NaN  102.0  274.0  0.372 NaN  NaN  NaN   
2  2.8  3.6    NaN   NaN   NaN  NaN   NaN  174.0  499.0  0.349 NaN  NaN  NaN   
3 -0.1 -0.6    NaN   NaN   NaN  NaN   NaN   22.0   86.0  0.256 NaN  NaN  NaN   
4 -0.1 -0.6    NaN   NaN   NaN  NaN   NaN   21.0   82.0  0.256 NaN  NaN  NaN   

      2P    2PA    2P%   eFG%     FT    FTA    FT%  ORB  DRB  TRB    AST  STL  \
0  144.0  516.0  0.279  0.279  170.0  241.0  0.705  NaN  NaN  NaN  176.0  NaN   
1  102.0  274.0  0.372  0.372   75.0  106.0  0.708  NaN  NaN  NaN  109.0  NaN   
2  174.0  499.0  0.349  0.349   90.0  129.0  0.698  NaN  NaN  NaN  140.0  NaN   
3   22.0   86.0  0.256  0.256   19.0   34.0  0.559  NaN  NaN  NaN   20.0  NaN   
4   21.0   82.0  0.256  0.256   17.0   31.0  0.548  NaN  NaN  NaN   20.0  NaN   

   BLK  TOV     PF    PTS  
0  NaN  NaN  217.0  458.0  
1  NaN  NaN   99.0  279.0  
2  NaN  NaN  192.0  438.0  
3  NaN  NaN   29.0   63.0  
4  NaN  NaN   27.0   59.0

A second file, players.csv, contains static information for each player, as height, weight, etc.

In [137]:
players = pd.read_csv('data/Players.csv', index_col=0)
players.head(5)

Player  height  weight                          collage    born  \
0  Curly Armstrong   180.0    77.0               Indiana University  1918.0   
1     Cliff Barker   188.0    83.0           University of Kentucky  1921.0   
2    Leo Barnhorst   193.0    86.0         University of Notre Dame  1924.0   
3       Ed Bartels   196.0    88.0  North Carolina State University  1925.0   
4      Ralph Beard   178.0    79.0           University of Kentucky  1927.0   

    birth_city birth_state  
0          NaN         NaN  
1     Yorktown     Indiana  
2          NaN         NaN  
3          NaN         NaN  
4  Hardinsburg    Kentucky

I then selected these three data categories:
    
    Player - Player's full name (first and last)
    height - Height in cm
    weight - Weight in kg

1.The players have unique names (checked at the beginning), we can merge the two dataframes using the 'Player' column.

In [138]:
data = pd.merge(df2, players[['Player', 'height', 'weight']], left_on='Player', right_on='Player', right_index=False,
      how='left', sort=False)
#.fillna(value=0)

2.Using rate statistics (i.e. points per game) or cumulative statistics (i.e. total points) can be misleading when it comes to analysis because these statistics tend to inflate players with lengthier careers. To deal with outliers, I instituted a minimum threshold of 40 games played. Also I keep only players with more than 400 minutes for each season (with a 82 games regular season, thats around 5 minutes per game. Players with less than that will be only anecdotical, and will distort the analysis).

In [139]:
data = data[~(data['Pos']==0) & (data['MP'] > 400) & (data['G'] > 40)]
data.reset_index(inplace=True, drop=True)

In [140]:
data.Pos.value_counts()

PF       3171
SF       3066
SG       3044
PG       2972
C        2960
SF-SG      22
C-PF       21
SG-SF      17
SG-PG      17
PG-SG      17
PF-C       16
PF-SF      15
SF-PF      13
F-C         9
G-F         6
SG-PF       3
G           3
C-F         3
F-G         2
C-SF        2
F           2
PG-SF       1
SF-PG       1
Name: Pos, dtype: int64

In [141]:
data = data[data.Pos.isin(['PF', 'PG', 'C', 'SG', 'SF'])]

3.Replace the * sign in some of the names.

In [142]:
data['Player'] = data['Player'].str.replace('*','')

4.For the stats that represent total values (others, as TS%, represent percentages), we will take the values per 36 minutes. The reason is to judge every player according to his characteristics, not the time he was on the floor.

In [143]:
totals = ['PER', 'OWS', 'DWS', 'WS', 'OBPM', 'DBPM', 'BPM', 'VORP', 'FG', 'FGA', '3P', '3PA', '2P', '2PA', 'FT', 'FTA',
         'ORB', 'DRB', 'TRB', 'AST', 'STL', 'BLK', 'TOV', 'PF', 'PTS']

for col in totals:
    data[col] = 36 * data[col] / data['MP']

In [144]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 15213 entries, 0 to 15382
Data columns (total 53 columns):
Unnamed: 0    15213 non-null int64
Year          15213 non-null float64
Player        15213 non-null object
Pos           15213 non-null object
Age           15211 non-null float64
Tm            15213 non-null object
G             15213 non-null float64
GS            11070 non-null float64
MP            15213 non-null float64
PER           15212 non-null float64
TS%           15213 non-null float64
3PAr          11543 non-null float64
FTr           15213 non-null float64
ORB%          12868 non-null float64
DRB%          12868 non-null float64
TRB%          13409 non-null float64
AST%          14117 non-null float64
STL%          12868 non-null float64
BLK%          12868 non-null float64
TOV%          12039 non-null float64
USG%          12039 non-null float64
OWS           15212 non-null float64
DWS           15212 non-null float64
WS            15212 non-null float64
WS/48   

from the data information above, the ideal number of data rows we want to work with is 15383, most amount of data is missing from 

    3P  (3-Point Field Goals)                  11543 non-null float64
    3PA (3-Point Field Goal Attempts)          11543 non-null float64
    3P% (3-Point Field Goal Percentage)        10478 non-null float64
    3PAr(3-Point Attempt Rate)                 11543 non-null float64
    
It's becasue 3-point line was not introduced to NBA until 1979, but for the sake of classifying modern players, I have to assume all players before 1979 have not attempted 3-pointers, so I will fill the missing values of these three columns with 0. 

In addition:

    GS  (Games Started)                        11070 non-null float64

Since number of Games started for each player does not effect their performance measuring, I will delete this column.

In [145]:
threePointsCol = ['3P','3PA','3P%','3PAr']
for feature in threePointsCol:
    data[feature] = data[feature].fillna(0)
del data['GS']
data[threePointsCol].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 15213 entries, 0 to 15382
Data columns (total 4 columns):
3P      15213 non-null float64
3PA     15213 non-null float64
3P%     15213 non-null float64
3PAr    15213 non-null float64
dtypes: float64(4)
memory usage: 594.3 KB


The next groups of columns that have substantical amount of missing data is from:
    
    ORB           12868 non-null float64
    DRB           12868 non-null float64
    STL           12868 non-null float64
    BLK           12868 non-null float64
    TOV           12039 non-null float64
    
And their associate percentages(plus USG%, relevant to TOV):
    
    ORB%          12868 non-null float64
    DRB%          12868 non-null float64
    STL%          12868 non-null float64
    BLK%          12868 non-null float64
    TOV%          12039 non-null float64
    USG%          12039 non-null float64
    
Since the above data has minimul 12039 or 12868 rows, the commonality is due to the fact ORB, DRB, STL, BLK and their associate percentages are not included in this dataset before 1974, and TOV data are not present until 1978. These missing values cannot be filled with assumption, but at the same time these columns are too important to be removed from the construction of player classification, therefore all the data before 1978 have to be discarded. 

In [146]:
data = data[~(data['Year'] < 1978)]

5.Cast these columns to be of type 'int'.

In [147]:
integerCol = ['Year','Age', 'G']
for feature in integerCol:
    data[feature] = data[feature].astype(dtype ='int')

In [148]:
pd.options.display.max_columns = None
data.sample(5).transpose()

13948           9222        12609          10036       5262 
Unnamed: 0       22253           14326       19954          15712        7934
Year              2014            1999        2010           2002        1986
Player      Quincy Acy  Olden Polynice  J.J. Barea  Allan Houston  Gene Banks
Pos                 SF               C          PG             SG          SF
Age                 23              34          25             30          26
Tm                 TOT             SEA         DAL            NYK         CHI
G                   63              48          78             77          82
MP                 847            1481        1546           2914        2139
PER            0.42928        0.286833     0.29806       0.187783     0.26087
TS%               0.52           0.461       0.526           0.54       0.559
3PAr             0.106           0.003       0.298          0.266       0.028
FTr              0.376           0.263       0.171          0.261       0.371
ORB%               9.5            14.2         1.3            1.5         9.1
DRB%              19.5              19         9.5            8.6          10
TRB%              14.4            16.6         5.4              5         9.5
AST%               4.9             4.7        26.1           12.6        16.6
STL%               1.4             0.7         1.2              1         1.8
BLK%               2.5             1.5         0.3            0.3         0.3
TOV%              15.4            10.9        15.3           10.5        14.8
USG%              10.1            13.7        19.9           25.9        17.9
OWS          0.0255018       0.0316003   0.0232859      0.0444749   0.0622721
DWS          0.0382527       0.0170155   0.0279431      0.0135896   0.0100982
WS           0.0637544       0.0486158    0.051229      0.0580645   0.0723703
WS/48            0.086           0.065       0.069          0.077       0.097
OBPM         -0.106257      -0.0413234  -0.0139715      0.0185312   0.0403927
DBPM         0.0595041      -0.0145847  -0.0419146     -0.0259437  -0.0117812
BPM         -0.0467532      -0.0559082  -0.0558862    -0.00864791   0.0302945
VORP        0.00850059     -0.00243079 -0.00232859      0.0123542   0.0336606
FG             2.80519         4.10804     5.40233        7.01716     5.99158
FGA            5.99292         8.70223     12.2717        16.0728     11.5792
FG%              0.468           0.472        0.44          0.437       0.517
3P            0.170012       0.0243079     1.30401        1.68016           0
3PA           0.637544       0.0243079     3.65589        4.27454    0.319776
3P%              0.267               1       0.357          0.393           0
2P             2.63518         4.08373     4.09832        5.33699     5.99158
2PA            5.35537         8.67792     8.61578        11.7982     11.2595
2P%              0.492           0.471       0.476          0.452       0.532
eFG%             0.482           0.473       0.493          0.489       0.517
FT              1.4876        0.704929     1.76973        3.64447     3.07994
FTA            2.25266         2.28494     2.09573        4.18806     4.29173
FT%               0.66           0.309       0.844           0.87       0.718
ORB            3.06021         4.47265     0.39586       0.457104     2.99579
DRB            6.12043          5.8582     3.00388        2.65614     3.06311
TRB            9.18064         10.3309     3.39974        3.11325     6.05891
AST            1.19008         1.04524     6.00776        2.34729      4.2244
STL           0.977568        0.486158    0.815006       0.667124     1.36325
BLK            1.10508        0.729237    0.139715       0.123542    0.168303
TOV            1.27509         1.19109     2.37516        2.10021     2.33941
PF             5.18536         3.64619     2.77102        2.24846     3.56802
PTS              7.268         8.94531     13.8784         19.359     15.0631
height             201             211 

In [149]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12039 entries, 3218 to 15382
Data columns (total 52 columns):
Unnamed: 0    12039 non-null int64
Year          12039 non-null int64
Player        12039 non-null object
Pos           12039 non-null object
Age           12039 non-null int64
Tm            12039 non-null object
G             12039 non-null int64
MP            12039 non-null float64
PER           12039 non-null float64
TS%           12039 non-null float64
3PAr          12039 non-null float64
FTr           12039 non-null float64
ORB%          12039 non-null float64
DRB%          12039 non-null float64
TRB%          12039 non-null float64
AST%          12039 non-null float64
STL%          12039 non-null float64
BLK%          12039 non-null float64
TOV%          12039 non-null float64
USG%          12039 non-null float64
OWS           12039 non-null float64
DWS           12039 non-null float64
WS            12039 non-null float64
WS/48         12039 non-null float64
OBPM       

In [150]:
data.to_csv('data/Seasons_Stats_cleansed.csv', sep='\t')